model/map/src/main/java/org/keycloak/models/map/realm/entity/MapAuthenticationExecutionEntity.java

```diff
@@ -0,0 +1,143 @@
+        entity.setAuthenticatorConfig(model.getAuthenticatorConfig());
+        entity.setFlowId(model.getFlowId());
+        entity.setParentFlowId(model.getParentFlow());
+        entity.setRequirement(model.getRequirement());
+        entity.setAutheticatorFlow(model.isAuthenticatorFlow());
+        entity.setPriority(model.getPriority());
+        return entity;
+    }
+
+    public static AuthenticationExecutionModel toModel(MapAuthenticationExecutionEntity entity) {
+        if (entity == null) return null;
+        AuthenticationExecutionModel model = new AuthenticationExecutionModel();
+        model.setId(entity.getId());
+        model.setAuthenticator(entity.getAuthenticator());
+        model.setAuthenticatorConfig(entity.getAuthenticatorConfig());
+        model.setFlowId(entity.getFlowId());
+        model.setParentFlow(entity.getParentFlowId());
+        model.setRequirement(entity.getRequirement());
+        model.setAuthenticatorFlow(entity.isAutheticatorFlow());
+        model.setPriority(entity.getPriority());
+        return model;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    public String getAuthenticator() {
+        return authenticator;
+    }
+
+    public void setAuthenticator(String authenticator) {
+        this.authenticator = authenticator;
+    }
+
+    public String getAuthenticatorConfig() {
+        return authenticatorConfig;
+    }
+
+    public void setAuthenticatorConfig(String authenticatorConfig) {
+        this.authenticatorConfig = authenticatorConfig;
+    }
+
+    public AuthenticationExecutionModel.Requirement getRequirement() {
+        return requirement;
+    }
+
+    public void setRequirement(AuthenticationExecutionModel.Requirement requirement) {
+        this.requirement = requirement;
+    }
+
+    public boolean isAutheticatorFlow() {
+        return autheticatorFlow;
+    }
+
+    public void setAutheticatorFlow(boolean autheticatorFlow) {
+        this.autheticatorFlow = autheticatorFlow;
+    }
+
+    public String getFlowId() {
+        return flowId;
+    }
+
+    public void setFlowId(String flowId) {
+        this.flowId = flowId;
+    }
+
+    public String getParentFlowId() {
+        return parentFlowId;
+    }
+
+    public void setParentFlowId(String parentFlowId) {
+        this.parentFlowId = parentFlowId;
+    }
+
+    public int getPriority() {
+        return priority;
+    }
+
+    public void setPriority(int priority) {
+        this.priority = priority;
+    }
+
+    @Override
+    public int hashCode() {
+        return getId().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (!(obj instanceof MapAuthenticationExecutionEntity)) return false;
+        final MapAuthenticationExecutionEntity other = (MapAuthenticationExecutionEntity) obj;
+        return Objects.equals(other.getId(), getId());
```

Comment:
Do we need this this equals implementation? I just realised it might be a problem (I checked AbstractRealmEntity and it is not the case currently), if we decide based on this equals whether to update RealmEntity in the storage or not. If we change, for example, `flowId` and we update the `updated` flag using `!Objects.equals(newVersion, oldVersion)` it will return false (a.k.a the objects are the same, we don't need to update storage). I just want to avoid future problems if we don't need this equals currently.

The solution would be to remove equals or compare all values. We should do the same also for other occurences.

Reply:
I believe two things are here: `updated` flag should be in each of the entities. And then what the `equals` is really used for.
This implementation is on par with similar implementations in e.g. `RealmAdapter`: https://github.com/keycloak/keycloak/blob/ca81e6ae8c5036581c21bbe5c4923d1af8fed55e/model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java#L1056-L1063
It captures the fact that the same entity can come from two sources, once e.g. from Infinispan, another time from JPA. Still, when maintaining set of the executions via e.g. a `HashSet`, this set should contain only one of the instances, regardless of the source, since the object identified is still the same.

Reply:
> Currently, I am not sure the updated flag makes sense on this entity. To me it seems we should use it only in case we take the nested entity and call some setter on it in adapter/provider. I haven't found anything like this for current entities.

The reason for having fine-grained updated flag is that then a provider which e.g. stores dependent entites in separate tables can optimize what really needs to be updated in the database.

> The way we introduced users nested entities is a little bit weird. We are combining two approaches (manually updating using an update method or using updated flag and update automatically) together and I think we should choose only one. The biggest advantage of the updated flag is that the code that works with users/clients/roles/etc. can just take model and call `setAnyProperty()` and it is projected in the storage on commit. However, with current nested entities this is somewhere in the middle. Entities support this, but models don't which is why we have all those `update*` methods.

This is a good catch. We should eventually get rid of the `update*` methods on models. As you write, this is something that could inflate the PR a lot if done here. I wonder though whether this removal is needed functionally? It seems to me that the more coarse-grained approach in this PR prepares ground for removal of the methods, and allows pushing removal of these methods and the need for them once map storage is implemented. This PR already proposes enough changes in my opinion.

> For example, this code: https://github.com/keycloak/keycloak/blob/c3b9c669418f7f4a55b5645ea0f3aa8321369345/services/src/main/java/org/keycloak/services/resources/admin/AuthenticationManagementResource.java#L1235-L1242
> 
> Currently:
> 
> 1. When we reach the code, we have RealmModel that contains entity that is registered in MapTransaction.
> 2. We call [this method](https://github.com/keycloak/keycloak/blob/da4c4c05e3009f4a699a9a3d1c989ddba34295f4/model/map/src/main/java/org/keycloak/models/map/realm/MapRealmAdapter.java#L735-L737) which calls `MapAuthenticatorConfigEntity.fromModel`  --> an instance of `MapAuthenticatorConfigEntity` is created with updated flag true
> 3. Then we call [updateAuthenticatorConfig](https://github.com/keycloak/keycloak/blob/da4c4c05e3009f4a699a9a3d1c989ddba34295f4/model/map/src/main/java/org/keycloak/models/map/realm/AbstractRealmEntity.java#L900-L902) that replaces the entity and always set updated flag to true if a config with the same id existed.
> 
> If we refactor it so that `AuthenticatorConfigModel` makes changes in the corresponding entity, we would change only one updated flag, update user in the storage only when `AuthenticatorConfigModel` is really changed, remove update methods from User/Realm adapters.

You are right, the `updated` is changed based on "the authenticator config has been touched" rather than it's been really different. An optimization for the future would be to go to refine `updated` flag to grasp that level of detail.

> If we decide to incorporate this, we should probably not do it in this PR as it would be huge then.

I 100% agree with this.

> Otherwise I would probably vote for removing the updated flag from entities where it is not necessary and leave the responsibility for setting updated flag on methods in AbstractUserEntity. This brings us back to the equals method above. Currently, using this method we cannot say whether the entity was updated or not because it checks only id. So we cannot do something like this when updating ExcutionEntity (but maybe we are ok with always updating):
> 
> ```
> MapAuthenticationExecutionEntity originalValue = authenticationExecutions.replace(authenticationExecution.getId(), authenticationExecution);
> this.updated |= originalValue != null && !Objects.equals(originalValue, authenticationExecution);
> ```

That is right. At this moment we're ok with always updating whenever there was any change (i.e. as coarse as calling of of the `update*` methods). It would allow getting the working realm map storage into the community to try and test, with ability to do the optimizations of the updates later.

Reply:
I agree, the code currently doesn't require the removal functionally. So we can leave this as is. I will go through the code an check whether the `equals` method is not used in the described way. 

I filed a Jira: https://issues.redhat.com/browse/KEYCLOAK-17626 @hmlnarik could you please triage? I am not sure about the correct fix version.